op-node/cmd/batch_decoder/fetch/fetch.go

```diff
@@ -32,45 +34,67 @@ type TransactionWithMetadata struct {
 }
 
 type Config struct {
-	Start, End   uint64
-	ChainID      *big.Int
-	BatchInbox   common.Address
-	BatchSenders map[common.Address]struct{}
-	OutDirectory string
+	Start, End         uint64
+	ChainID            *big.Int
+	BatchInbox         common.Address
+	BatchSenders       map[common.Address]struct{}
+	OutDirectory       string
+	ConcurrentRequests uint64
 }
 
 // Batches fetches & stores all transactions sent to the batch inbox address in
 // the given block range (inclusive to exclusive).
 // The transactions & metadata are written to the out directory.
-func Batches(client *ethclient.Client, config Config) (totalValid, totalInvalid int) {
+func Batches(client *ethclient.Client, config Config) (totalValid, totalInvalid uint64) {
 	if err := os.MkdirAll(config.OutDirectory, 0750); err != nil {
 		log.Fatal(err)
 	}
-	number := new(big.Int).SetUint64(config.Start)
 	signer := types.LatestSignerForChainID(config.ChainID)
+	concurrentRequests := int(config.ConcurrentRequests)
+
+	g, ctx := errgroup.WithContext(context.Background())
+	g.SetLimit(concurrentRequests)
+
+loop:
 	for i := config.Start; i < config.End; i++ {
-		valid, invalid := fetchBatchesPerBlock(client, number, signer, config)
-		totalValid += valid
-		totalInvalid += invalid
-		number = number.Add(number, common.Big1)
+		select {
+		case <-ctx.Done():
+			break loop
+		default:
+		}
+		number := i
+		g.Go(func() error {
+			valid, invalid, err := fetchBatchesPerBlock(client, number, signer, config)
+			if err != nil {
+				return fmt.Errorf("error occurred while fetching block %d: %w", number, err)
+			}
+			atomic.AddUint64(&totalValid, valid)
+			atomic.AddUint64(&totalInvalid, invalid)
+			return nil
+		})
+	}
+	if err := g.Wait(); err != nil {
+		log.Fatal(err)
 	}
 	return
 }
 
 // fetchBatchesPerBlock gets a block & the parses all of the transactions in the block.
-func fetchBatchesPerBlock(client *ethclient.Client, number *big.Int, signer types.Signer, config Config) (validBatchCount, invalidBatchCount int) {
-	ctx, cancel := context.WithTimeout(context.Background(), 3*time.Second)
+func fetchBatchesPerBlock(client *ethclient.Client, number uint64, signer types.Signer, config Config) (uint64, uint64, error) {
```

Comment:
```suggestion
func fetchBatchesPerBlock(ctx context.Context, client *ethclient.Client, number uint64, signer types.Signer, config Config) (uint64, uint64, error) {
```

fetchBatchesPerBlock should take a context as the first parameter.